release/src/router/totd/totd.8
.TH "totd" "8" "1.5" "Feike W. Dillema" "net"
.SH "NAME"
.LP 
totd \- DNS proxy and translator for IPv6 and IPv4
.SH "SYNTAX"
.LP 
totd [\fI\-6\fP] [\fI\-no6\fP] [\fI\-4\fP] [\fI\-no4\fP] [\fI\-64\fP] [\fI\-d<n>\fP] [\fI\-p\fP <prefix>] [\fI\-c\fP <configfile>]  [\fI\-u\fP <user>]  [\fI\-g\fP <group>]  [\fI\-t\fP <dir>] [\fI\-q\fP] [\fI\-v\fP] [\fI\-h\fP]
.SH "DESCRIPTION"
.LP 
Totd is a small DNS proxy nameserver. Its main purpose is to support IPv6\-only hosts/networks that communicate with the IPv4 world using some network or transport level translation mechanism, like NAT\-PT or faithd(8). In addition, it has some other useful tricks and features one might fancy in a DNS proxy.
Note that totd needs to be able to forward requests to a real nameserver (called a forwarder), in order to be useful.
.SH "CONFIGURATION"
.LP 
Totd can be configured using commandline options, via its configuration file or even dynamically via HTTP requests (only
if support for this has been added at compile time).
.LP 
If you want to run totd on a privileged port, e.g. the default port 53, you either have to run totd as root or start as root but let it switch to another user and group after it has opened the sockets its needs (see -u and -g options). In the former case, totd is able to rescan the interface list and deal with interface and address changes on the fly. In the latter case, totd will not be able to rescan interfaces and react to such changes but needs to be restarted to handle such changes.
.LP 
In the totd config file the first word is a keyword and it is followed by a value and or one or more attributes or optional values. Stuff between square brackets `[' and `]' below is optional while stuff between hooks`<' and `>' signifies a value you supply. (Neither the square brackets nor hooks are meant to be in the config file).
.LP 
The following keywords are valid in the totd config file:
.TP 
\fBforwarder\fR <ip address> [port <service port>] 
.br 
With this keyword you can specify an IP address (either IPv4 or IPv6) of a nameserver for totd to forward its queries to.
With the optional port attribute an alternative port can be specified to query the forwarder on.  The default is the standard domain service port, i.e. 53. This feature is mainly useful for running totd and a forwarder nameserver, e.g. bind on the same machine. In such a case, clients can talk to totd on port 53, totd talks to named over some other port.
.br 
At least one forwarder needs to be specified in order for totd to have non\-trivial behaviour.
.TP 
\fBprefix\fR <IPv6 network prefix>
.br 
Specifies a 64 bit IPv6 address prefix to use for the `address translation trick' described in more detail below.
The prefix must be written in IPv6 address format like (but without the quotes): `3ffe:1234:abcd:1234::'.
.TP 
\fBallow\fR <IP addresstring>
.br 
Specifies an (IPv4 or IPv6) address of a machine that is allowed to make http requests to totd.
.TP 
\fBport\fR <port number>
.br 
This option allows you to specify a different (TCP/UDP) port than the default (53) for totd to listen on.
.TP 
\fBinterfaces\fR <ifa> [<ifb> <ifc> ...]
.br 
By default totd listens on wildcard sockets. This option allows you to specify what network interfaces totd should listen on for incoming requests. Note, that his option is not available on all operating systems (not on Linux e.g.). When '*' is named as interface, the the interface list is ignored and wildcard sockets are used. If 'all' is named as interface, totd will scan all interfaces for addresses and open a socket to listen on for each of them. (Mostly useful in combination with scoped address rewriting, see scoped keyword).
.TP 
\fBpidfile\fR <filename>
.br 
Changes the default filename for totd to write its process id into.
\fBstf\fR 
.br 
Enable 6to4 PTR rewriting.
.TP 
\fBscoped\fR <v6 prefix 1> <v6 prefix 2> <prefixlen>
.br 
Enable scoped address rewriting. This requires currently that the interfaces keyword is also used, i.e. scoped address rewriting is not supported for wildcard sockets.




.SH "OPTIONS"
.LP 
.TP 
\fB\-d\fR<\fIn\fP>
Run totd in foreground to facilitate debugging and set debugging output level to \fIn\fP. 
.TP 
\fB\-6\fR
Listen on IPv6 addresses for client queries
.TP 
\fB\-no6\fR
Do not listen on IPv6 addresses
.TP 
\fB\-4\fR
Listen on IPv4 addresses
.TP 
\fB\-no4\fR
Do not listen on IPv4 addresses
.TP 
\fB\-64\fR
Listen to both IPv4 and IPv6 addresses
.TP 
\fB\-p\fR <\fIprefix\fP>
Add a IPv6 prefix to trick list
.TP 
\fBhttp-port\fR <port number>
.br 
Specifies a different port than the default (6464) to listen on for http requests. Note that http support is optional and needs to be compiled in.
.TP 
\fB\-c\fR <\fIfilename\fP>
Specify alternative config file for totd to read in
.TP 
\fB\-u\fR <\fIuser\fP>
Specify username or uid totd should switch to
.TP 
\fB\-g\fR <\fIgroup\fP>
Specify groupname or gid totd should switch to
.TP 
\fB\-t\fR <\fIdir\fP>
Specify a root directory to chroot() to
.TP 
\fB\-h\fR
Output version and usage information and exit
.TP 
\fB\-q\fR
Log only errors to syslog, be more quiet.
.TP 
\fB\-v\fR
Log more verbosely, also log warnings and informational messages.
.LP 
.SH "HTTP REQUESTS"
.LP 
When support for it is compiled into totd and totd is started with the http-port commandline
option, the following URLs can be requested from totd (at the specified port, say 6464):
.LP 
http://localhost:6464/index.html
.br 
http://localhost:6464/add_prefix?prefix=3ffe:abcd:1234:9876::
.br 
http://localhost:6464/del_prefix?prefix=3ffe:abcd:1234:9876::
.LP 
Of course, you choose your own values for the prefix to add or delete. Note that if your network
can not be considered secure, it may be appropriate to let totd only listen to localhost for
http requests (and use ssh-tunneling for remote access)? Basic http authentication could easily
be supported, but we do not see much value in that. HTTPS support is more complicated and I
currently do not believe it adds much over using ssh.
.LP 
.SH "FORWARDING"
.LP 
Totd is just a DNS\-proxy; it does not answer queries itself. It can only forward queries to a real nameserver. If there are multiple forwarders specified, it will use them in the order listed.  When a nameserver is or becomes unreachable totd will use the next nameserver in line. After the retry interval amount of time, totd will switch back to the previous nameserver. If that nameserver is still unreachable it uses the next nameserver in the config file again. You may see totd switch to backup nameservers for no apparent good reason, as totd is not very good at discriminating between an unreachable/malfunctioning nameserver or a single query that returns erronous results or is simply delayed a lot. 
.br 
(If the second nameserver is also unreachable and a third is specified, this sequence repeats itself.)
.SH "ADDRESS TRANSLATION TRICK"
.LP 
Totd can treat each AAAA and A6 type query in a special way. This behaviour is enabled when one or more prefixes are configured (on the commandline or with the `prefix' keyword in the config file). It is meant to support network and transport level IPv6 to IPv4 transition mechanisms, like NAT\-PT and faith.
.br 
If the nameserver does not return an IPv6 address for the forwarded AAAA/A6 query, totd will make a second query but this time for an A record of the hostname of the original query. The resulting IPv4 address is then used to construct a fake IPv6 address, by replacing the lower 32 bits of the specified prefix with this IPv4 address. The resulting IPv6 address is sent as response to the original AAAA/A6 record query. 
.br 
In addition, totd treats PTR type queries (reverse name lookup) in the ip6.int. domain specially. If the query matches a specified prefix, totd will forward a PTR query for an IPv4 address (using the lower 32 bits of the address to construct an IPv4 address) instead and use that to construct a faked response to the original PTR query.
.br 
If multiple prefixed are configured, totd will cycle through them in round\-robin fashion. In this way totd can balance the load for multiple NAT\-PT/faithd(8) translators in a network.
.SH "6TO4 PTR RECORDS"
.LP 
To enable 6to4 PTR query translation trick, add the `stf' keyword to your totd config file. 6to4 reverse lookup is based on draft\-moore\-6to4\-dns\-00.txt, section 3.3:
.IP 
     When such a resolver received a PTR or NS query for a label that had a [x2002].IP 6.ARPA suffix, it would first attempt to satisfy that query from its cache, or failing that, by forwarding the query to an upstream server.  If that query failed due to a "no such domain" error,the resolver would then attempt to find the server for the {something}.[x2002].IP 6.ARPA label by issuing an NS query for {something}.I N\-ADDR.ARPA.
.IP 
     If the original query was for PTR records, and one or more NS records were found for {something}.I N\-ADDR.ARPA, the resolver would then forward the original query for {something}.[x2002].IP 6.ARPA to one or more of those servers, and return the results from one of the forwarded queries if any were successful.
.IP 
     If the original query was for NS records, and one or more NS records were found for {something}.I N\-ADDR.ARPA, the resolver would then return the pseudo\-records corresponding to the IN\-ADDR.ARPA domains. Those pseudo\-records would NOT be marked as authoritative, and the resolver would NOT cache those records.
.SH "SCOPED ADDRESSES"
.LP 
Totd supports re\-writing of scoped addresses in DNS responses.
This technique allows usage of scoped addresses like site\-local
addresses, without having to maintain  such addresses in a DNS database. Instead, totd derives the scoped address record from the global address record it does find in DNS. 
.LP 
Totd only performas this trick on queries that stay within the specified scope. I.e. if the query is made from scoped source address (link\-local unicast or site\-local unicast), and query's target address (totd's listening address) is also scoped address, totd attaches additional AAAA records converted by using 3 arguments of the scoped keyword.
.LP 
When you configure as below in totd config file:
.TP 
\fBscoped\fR 3ffe:501:ffff:: fec0:: 48

.LP 
and you made query from scoped source to totd's scoped destination, and the result has the following record:

.TP 
foo.kame.net.   \fBIN AAAA\fR 3ffe:501:ffff::9876:5432
.LP 
it will get additional records as follows:
.TP 
foo.kame.net.   \fBIN AAAA\fR 3ffe:501:ffff::9876:5432
.TP 
foo.kame.net.   \fBIN AAAA\fR fec0::9876:5432
.TP 
At this moment, prefixlen must be multiple of 8.
.LP 
Reverse query for fec0::9876:5432 will be converted into 3ffe:501:ffff::9876:5432 and forwarded to the real DNS servers.
.LP 
.SH "EXAMPLES"
.LP 
The Kame project provided network (NAT\-PT) and transport (faithd) level translators for the *BSD operating systems, see e.g. http://www.kame.net for more information.
.br 
Another transport level translator is 'The Portable Transport Relay Translator Daemon (pTRTd)' which supports Linux amongst others, see
http://v6web.litech.org/ptrtd/
.LP 
.SH "FILES"
.LP 
\fIPREFIX/etc/totd.conf\fP
.SH "AUTHOR"
.LP 
Feike W. Dillema <feico at pasta.cs.uit.no>

Thanks to 6Net (EU project IST\-2001\-32603), http://www.6net.org/
.SH "SEE ALSO"
.LP 
faithd(8) natptd(8) natptconfig(8) stf(4)


release/src/router/totd/totd.c
/****************************************************************************
 * Copyright (C) 1998 WIDE Project. All rights reserved.
 * Copyright (C) 1999,2000,2001,2002 University of Tromso. All rights reserved.
 * Copyright (C) 2002 Invenia Innovation AS. All rights reserved.
 *
 * Author: Feike W. Dillema, feico@pasta.cs.uit.no.
 *         based on newbie code by Yusuke DOI, Keio Univ. Murai Lab.
 ****************************************************************************/

/*
 * <$Id: totd.c,v 3.74 2005/02/02 11:10:31 dillema Exp $>
 */

#include "totd.h"

struct ToT T;
char *version = "Trick or Treat Daemon (totd) version 1.5";

void usage () {
	printf ("%s\n\
Usage: totd [-6|-no6|-4|-no4|-64|-dn|-v|-q|-p <prefix>|-c <filename>]\n\
\n\
-[no]6      : enable[disable] IPv6 service functionality\n\
-[no]4      : enable[disable] IPv4 service functionality\n\
-64         : alias to -6 -4\n\
-dn         : debug mode (no fork / loglevel = n)\n\
-v          : verbose\n\
-q          : quiet\n\
-u <user>   : username or uid totd should run at, after startup\n\
-g <group>  : groupname or gid totd should run at, after startup\n\
-t <dir>    : put totd in <dir> chroot() cage\n\
-p <prefix> : a prefix to use for totd tricks; more than one allowed\n\
-http-port <port> : port we listen on for http requests (default = 6464)\n\
-c <file>   : specify alternative totd configfile, default=%s\n\
\n\
default   : IPv6 is %s and IPv4 is %s\n\n\
totd %s use IPv6 because it was compiled %s USE_INET6 option in config.h\n\
totd %s use IPv4 because it was compiled %s USE_INET4 option in config.h\n",
		version,
		TOTCONF,
		(T.ip6) ? "enabled" : "disabled",
		(T.ip4) ? "enabled" : "disabled",
		(V4 (1) + 0) ? "can" : "can not",
		(V4 (1) + 0) ? "with" : "without",
		(V6 (1) + 0) ? "can" : "can not",
		(V6 (1) + 0) ? "with" : "without"
	);

	totd_exit (EXIT_FAILURE);
}

const char *hex = "0123456789abcdef";

int main (int argc, char **argv) {
	struct passwd *pwd_p;
	int i;

	/* initialize global totd structure */
	T.uid = getuid();
	T.gid = getgid();

	T.user = NULL;
	T.group = NULL;
	T.rootdir = NULL;

	/* fill in default config values */
	T.ip4 = V4 (1) + 0;	/* bit ugly, but short */
	T.ip6 = V6 (1) + 0;
	T.use_mapped = 0;
	T.quiet = 0;
	T.debug = 0;
	T.prefixnum = 0;
	T.rescan_iflist = 1;
	T.stf = 0;
	T.retry_interval = 300;
	T.port = PORT_SRV;
	T.http_port = 0;
	T.wildcard = 1;
	T.Fwd_list = NULL;
	T.pidfile = TOT_PID_FILE;
	T.configfile = TOTCONF;
	T.current_fwd = NULL;

	/* make sure these start out empty */
	for (i = 0; i < MAXPREFIXES; i++) {
		T.prefix[i][0] = '\0';
	}

	/* list of forwarders */
	T.Fwd_list = list_init ();
	if (!T.Fwd_list)
		exit(1);

	/* parse command line arguments */
	for (i = 1; i < argc; i++) {
		if (!strncmp (argv[i], "-d", 2) ) {
			/* debug option */
			T.debug = atoi(&argv[i][2]);
			syslog (LOG_INFO, "debug level %d enabled", T.debug);
		} else if (!strcmp (argv[i], "-6"))
			T.ip6 = 1;
		else if (!strcmp (argv[i], "-4"))
			T.ip4 = 1;
		else if (!strcmp (argv[i], "-46") || !strcmp (argv[i], "-64"))
			T.ip4 = 1, T.ip6 = 1;
		else if (!strcmp (argv[i], "-no4"))
			T.ip4 = 0;
		else if (!strcmp (argv[i], "-no6"))
			T.ip6 = 0;
		else if (!strcmp (argv[i], "-c"))
			T.configfile = strdup(argv[++i]);
		else if (!strcmp (argv[i], "-p")) {
			if (conv_trick_conf((u_char *)argv[++i])) {
			    syslog (LOG_ERR, "invalid prefix on command line: %s", argv[++i]);
			    usage(1);
			}
		}
		else if (!strcmp (argv[i], "-http-port")) {
			T.http_port = atoi(argv[++i]);
			if (!T.http_port) {
				syslog (LOG_ERR, "invalid portnumer: %s", argv[++i]);
				usage(1);
			}
		} else if (!strcmp (argv[i], "-u"))
			T.user = strdup(argv[++i]);
		else if (!strcmp (argv[i], "-g"))
			T.group = strdup(argv[++i]);
		else if (!strcmp (argv[i], "-t"))
			T.rootdir = strdup(argv[++i]);
		else if (!strcmp (argv[i], "-h"))
			usage (1);
		else if (!strcmp (argv[i], "-v"))
			T.quiet = -1;
		else if (!strcmp (argv[i], "-q"))
			T.quiet = 1;
		else {
			syslog (LOG_ERR, "unknown option %s", argv[i]);
			usage (1);
		}
	}

#ifndef LOG_PERROR
#define LOG_PERROR 0
#endif /* not LOG_PERROR */

	if (T.debug > 0)
	    openlog("totd", LOG_PID | LOG_NDELAY | LOG_CONS | LOG_PERROR, LOG_DAEMON);
	else
	    openlog ("totd", LOG_PID | LOG_NDELAY | LOG_CONS, LOG_DAEMON);

	syslog (LOG_NOTICE, "%s", version);

	if (T.debug) {
		if (T.quiet > 0)
			setlogmask (LOG_UPTO (LOG_ERR));
		else if (!T.quiet)
		setlogmask (LOG_UPTO (LOG_INFO));
		else
			setlogmask (LOG_UPTO (LOG_DEBUG));
	} else {
		if (T.quiet > 0)
			setlogmask (LOG_UPTO (LOG_ERR));
		else if (!T.quiet)
			setlogmask (LOG_UPTO (LOG_WARNING));
		else if (T.quiet < 0)
	       		setlogmask (LOG_UPTO (LOG_INFO));
	}

	if (T.user) {
                if (isdigit(T.user[0])) {
			T.uid = atoi(T.user);
			pwd_p = NULL;
		} else {
			pwd_p = getpwnam (T.user);
		}
	} else
		pwd_p = getpwuid (T.uid);

	if (pwd_p) {
		syslog (LOG_INFO, "Found user record of %s; uid: %d gid: %d",
			pwd_p->pw_name, pwd_p->pw_uid, pwd_p->pw_gid);
		if (T.uid && T.uid != pwd_p->pw_uid) {
			syslog (LOG_ERR, "Need root privileges to change user \
to: %s", T.user);
			totd_exit (EXIT_FAILURE);
		} else {
			T.uid = pwd_p->pw_uid;
		}
	} else {
		syslog (LOG_INFO, "can't find user record of %s", T.user);
	}

	if (T.group) {
		if (isdigit(T.group[0])) {
			T.gid = atoi(T.group);
		} else {
        		struct group *grp_p;

			grp_p = getgrnam(T.group);
			if (!grp_p) {
				syslog(LOG_ERR, "group `%s' unknown", T.group);
				totd_exit (EXIT_FAILURE);
			}
			T.gid = grp_p->gr_gid;
		}
	} else {
		if (pwd_p)
			T.gid = pwd_p->pw_gid;
		else
			T.gid = getgid();
	}

	/* close any open descriptors */
	endpwent();
	//endgrent();

	/* chroot() cage totd if asked for */
	if (T.rootdir) {
                if (chroot(T.rootdir) < 0) {
                        syslog (LOG_ERR, "chroot %s failed: %m", T.rootdir);
                        totd_exit(EXIT_FAILURE);
                } 
                syslog (LOG_INFO, "chrooted to %s", T.rootdir);
                if (chdir("/") < 0) {
                        syslog (LOG_ERR, "chdir(\"/\") failed: %m");
                        totd_exit(EXIT_FAILURE); 
                }
	}

#ifdef SWILL
  	/* Initialize the SWILL server */
	if (T.http_port) {
  		swill_init(T.http_port);
		syslog (LOG_INFO, "Listening on [*]:%d for http requests", T.http_port);
		swill_deny("");
	}
#endif
	/*
	 * read in config file
	 * possibly overriding defaults and command line options
 	 */
	if (read_config (T.configfile)) {
	    syslog (LOG_ERR, "Configuration failure");
	    syslog (LOG_INFO, "Check log, or try -d option for debug mode...");
	    totd_exit (EXIT_FAILURE);
	}

	/* we cannot rebind to reserved ports if we are not root */
	if (T.rescan_iflist == 1 && T.uid && T.port < IPPORT_RESERVED) {
		syslog(LOG_WARNING, "Disabling rescanning of network interfaces");
		T.rescan_iflist = 0;
	}

	/* Check for pidfile conflicts */
	if (T.pidfile) {
		FILE *pid_fp;
		pid_t pid;

		/* check if pid file exists */
		pid_fp = fopen (T.pidfile, "r");
		if (pid_fp) {
			if (fscanf (pid_fp, "%d", &pid) != 1) {
				syslog (LOG_NOTICE, "Removing bogus lockfile");
				unlink (T.pidfile);
			} else if (kill (pid, 0) == -1 && !unlink (T.pidfile)) {
				syslog (LOG_NOTICE, "Removed stale lockfile");
			} else {
				syslog (LOG_ERR, "PID file %s already exists",
					T.pidfile);
				syslog (LOG_INFO, "There can be another totd \
running. Please make sure there's no totd already running. Then delete the \
file %s and try again.", T.pidfile);
				fclose (pid_fp);
				totd_exit (EXIT_FAILURE);
			}
			fclose (pid_fp);
		}
	}

	if (!T.ip4 && T.ip6) {
		/* If user specifies we should *not* accept IPv4
		 * requests on Linux, we will bail out with an error.
		 * totd does not filter incoming requests on their source
		 * address. It is not totd's task to do so IMHO, and I will
		 * probably never implement that.
		 */
#ifdef NEEDSV4MAPPED
		syslog(LOG_ERR, "Cannot disable IPv4 when IPv6 is enabled on\
this OS, due to IPv4 mapped addresses");
		syslog(LOG_INFO, "Will always accept IPv4 *and* IPv6 requests.\
Bailing out, so that you can explicitly tell me to do so");
		totd_exit (EXIT_FAILURE);
#endif
	}

	if (T.ip4 && T.ip6) {
#ifdef WILDCARDONLY
		if (!T.wildcard) {
			syslog(LOG_ERR, "On this OS we only support \
wildcard binding when IPv6 is enabled.");
			syslog(LOG_INFO, "Please remove `interfaces' \
specification in your config file.");
			totd_exit (EXIT_FAILURE);
		}
#endif
#ifdef NEEDSV4MAPPED
		syslog(LOG_DEBUG, "IPv6 wildcard socket with IPv4 mapped, \
will not bind to wildcard IPv4 socket.");
		T.use_mapped = 1;
#endif
	}

	if (T.ip4) {
#ifdef USE_INET4
		syslog (LOG_INFO, "IPv4 activated");
#else
		syslog (LOG_ERR, "IPv4 support is not compiled in");
		totd_exit (EXIT_FAILURE);
#endif
	}

	if (T.ip6) {
#ifdef USE_INET6
		syslog (LOG_INFO, "IPv6 activated");
#else
		syslog (LOG_ERR, "IPv6 support is not compiled in");
		totd_exit (EXIT_FAILURE);
#endif
	}

	if (!T.ip4 && !T.ip6) {
		syslog (LOG_ERR, "all supported protocols are deactivated; \
what do you want me to do then?");
		totd_exit (EXIT_FAILURE);
	}

#ifdef SCOPED_REWRITE
	if (T.wildcard && T.scoped_prefixes) {
		syslog (LOG_ERR, "Scoped address rewriting currently not \
implemented when wildcard sockets are used. Please use `interfaces' keyword \
in your config file or remove `scoped' keyword");
		totd_exit (EXIT_FAILURE);
	}
#endif

	fwd_init();
	fwd_select();
	if (!T.current_fwd) {
	    syslog (LOG_ERR, "no forwarder available, what should we do then?");
	    return -1;
	}

	/* initialize each event routine */
	ev_dup_init ();
	if (ev_signal_init () < 0) {
		syslog (LOG_ERR, "Signal event handling  initialize failed");
		totd_exit (EXIT_FAILURE);
	}

	if (ev_to_init () < 0) {
		syslog (LOG_ERR, "Timeout event handling initialize failed");
		totd_exit (EXIT_FAILURE);
	}

	if (ev_tcp_conn_in_init () < 0) {
		syslog (LOG_ERR, "TCP connection initialize failed");
		totd_exit (EXIT_FAILURE);
	}

	if (net_init_socketlist(T.port) < 0) {
		syslog (LOG_ERR, "Init list of sockets failed");
		totd_exit (EXIT_FAILURE);
	}

	if (net_bind_socketlist() <= 0) {
		if (!T.rescan_iflist) {
			syslog (LOG_ERR, "Could not open any sockets");
			totd_exit (EXIT_FAILURE);
		} else {
			syslog (LOG_WARNING, "Could not open any sockets");
			syslog (LOG_WARNING, "Maybe later??? Continuing");
		}
	}

	if (ev_udp_in_init () < 0) {
		syslog (LOG_ERR, "UDP initialize failed");
		totd_exit (EXIT_FAILURE);
	}
	if (ev_tcp_out_init () < 0) {
		syslog (LOG_ERR, "TCP output routine initialize failed");
		totd_exit (EXIT_FAILURE);
	}

	/* drop root privs */
	if (setgid(T.gid) < 0) {
		syslog (LOG_ERR, "setgid to %d failed", T.gid);
		totd_exit (EXIT_FAILURE);
	}
	if (setuid(T.uid) < 0) {
		syslog (LOG_ERR, "setuid to %d failed", T.uid);
		totd_exit (EXIT_FAILURE);
	}

	if (T.rescan_iflist) {
		if (ev_to_register_ifcheck () < 0) {
			syslog (LOG_ERR, "Registering Interface Check Event failed");
			totd_exit (EXIT_FAILURE);
		}
	}

	if (!T.debug) {
		if (daemon(0,0))
			totd_exit (EXIT_FAILURE);
		else
			syslog (LOG_INFO, "totd successfully daemonized");
	}

	/* as detached child, we can now write pid file */
	if (T.pidfile) {
		FILE *pid_fp;
		pid_fp = fopen (T.pidfile, "w");
		if (!pid_fp) {
			syslog (LOG_ERR, "can't open pid file \"%s\"", T.pidfile);
			totd_exit (EXIT_FAILURE);
		}
		fprintf (pid_fp, "%d", getpid ());
		syslog (LOG_INFO, "wrote pid %d to file %s", getpid (), T.pidfile);
		fclose (pid_fp);
	}

#ifdef SWILL
	if (T.http_port) {
		if (T.debug) swill_log(stderr);
		swill_title("Trick or Treat DNS Proxy");
        	swill_handle("index.html", print_stats, 0);
        	swill_handle("add-prefix.html", add_prefix, 0);
        	swill_handle("del-prefix.html", del_prefix, 0);
	}
#endif
	syslog (LOG_INFO, "totd started");
	totd_eventloop ();
	return (totd_exit (0));
}

void totd_eventloop (void) {
	const char *fn = "totd_eventloop()";
	struct timeval tv_out, *tvp;
	fd_set fd_read, fd_write;
	int max_fd, fdnum, i;
	time_t next_timeout;
#ifdef DBMALLOC
	unsigned long histid1, histid2, orig_size, current_size;

	/* Malloc debugging */
	if (T.debug) 
		malloc_dump (2);
	orig_size = malloc_inuse (&histid1);
	syslog (LOG_DEBUG, "Malloc Size: %ld", orig_size);
#endif

	while (1) {		/* main loop */
		if (T.debug > 2)
			syslog (LOG_DEBUG, "main loop: start");

		/* pick a proper forwarder */
		fwd_select ();

#ifdef DBMALLOC
		/* Malloc debugging */
		current_size = malloc_inuse (&histid2);

		if (current_size != orig_size) {
			syslog (LOG_DEBUG, "Malloc Size: %ld", current_size);
			if (T.debug) 
				malloc_list (2, histid1, histid2);
			orig_size = current_size;
			/* histid1 = histid2; */
		}
#endif

#ifdef SWILL
		if (T.http_port) swill_poll();
#endif
		/* signal event */
		ev_signal_process ();

		/* timeout event */
#ifdef SWILL
		if (T.http_port) {
			tv_out.tv_usec = 500;
			tv_out.tv_sec = 0;
			tvp = &tv_out;
		} else
#endif
		{
			next_timeout = ev_timeout_process ();
			if (!next_timeout) {
				if (T.debug > 2)
					syslog (LOG_DEBUG, "no timeouts at present");
				tvp = NULL;
			} else {
				tv_out.tv_usec = 0;
				tv_out.tv_sec = next_timeout - time (NULL);
				tvp = &tv_out;
				if (T.debug > 2)
					syslog (LOG_DEBUG, "next timeout after %ld s.",
						tv_out.tv_sec);
			}
		}

		/* get FD_SET for now */
		max_fd = 0;
		FD_ZERO (&fd_read);
		FD_ZERO (&fd_write);

		if (T.debug > 3)
			syslog (LOG_DEBUG, "check for UDP fds...");

		nia_fds_set (&fd_read, &max_fd);

		if (T.debug > 3)
			syslog (LOG_DEBUG, "check for TCP-in fds...");
		i = ev_tcp_conn_in_fds (&fd_read);
		max_fd = MAXNUM (i, max_fd);

		if (T.debug > 3)
			syslog (LOG_DEBUG, "check for TCP-out fds...");
		i = ev_tcp_out_fds (&fd_write);
		max_fd = MAXNUM (i, max_fd);

		if (T.debug > 3)
			syslog (LOG_DEBUG, "%s: max_fd = %d", fn, max_fd);

		/* select */
		if (T.debug > 2)
			syslog (LOG_DEBUG, "main loop: select");

		fdnum = select (max_fd + 1, &fd_read, &fd_write, NULL, tvp);
		if (fdnum < 0) {
			if (errno == EINTR) {
				syslog (LOG_DEBUG, "%s: select() interrupted",
					fn);
				continue;	/* while(1) */
			} else {
				syslog (LOG_ERR, "%s: select(): %m", fn);
				if (net_reinit_socketlist (T.port, 1) < 0)
					totd_exit (EXIT_FAILURE);
				sleep (1);
			}
		} else {
			int sock;

			switch (nia_fds_isset (&fd_read, &sock)) {
			case 0: /* UDP */
				if (ev_udp_in_read (sock) < 0)
					syslog (LOG_INFO, "udp service error");
				continue;
			case 1: /* TCP */
				if (ev_tcp_srv_accept (sock) < 0)
					syslog (LOG_INFO, "tcp service error");
				continue;
			default: /* not found */
				if (ev_tcp_out_fd_check (&fd_write) < 0)
					syslog (LOG_INFO, "tcp output failed");

				if (ev_tcp_conn_in_fd_check (&fd_read) < 0)
					syslog (LOG_INFO, "tcp input failed");
			}
		}		/* if(...select...) */
	}			/* while(1) */
}

int totd_exit (int status) {

	/* finish in reverse order of initialize */
	ev_tcp_out_finish ();
	ev_udp_in_finish ();
	ev_tcp_conn_in_finish ();
	/* ev_tcp_srv_in_finish() -- no such func */
	ev_to_finish ();
	ev_signal_finish ();
        if (T.Fwd_list)
                list_destroy (T.Fwd_list, fwd_freev);

	if (T.pidfile)
		unlink (T.pidfile);

	if (status != EXIT_SUCCESS) {
		syslog (LOG_ERR, "terminated with error");
		fprintf (stderr, "totd terminated with error, \
check system logs for details or run totd in debug mode.\n");
	}

	exit (status);
}
